### ISSUE TITLE
Recursive macros work only when defined at some special places

### ISSUE THREAD

**Issue by benallard** [pallets/jinja#361]
I've had some troubles when the inner call of a recusrive macro (the first call to itself) threw a `UndefinedError`. The issue could be fixed by defining the macro at a higher level.

The following template should illustrate my issue:

``` jinja
{% set n=[1,2,3,4,5] %}
{% for n in [[1,2,3], [3,4,5], [5,6,7]] %}

{% macro x(l) %}
  {{ l.pop() }}
  {% if l %}{{ x(l) }}{% endif %}
{% endmacro %}

{{ x(n) }}

{% endfor %} 
```

If the `for` `endfor` block is commented out, it works fine, same if the macro is defined outside of the `for` block. (the `set` is here so that commenting out the `for` delimiter work.)

I've had the same trouble with other block as `for` (namely `block`, but can't reproduce in such a small example)

Using version 2.6.1 on Debian Wheezy.

**Comment by malthejorgensen** [comment]
+1 :+1: This should be noted in the docs (that you cannot define a recursive macro inside a `block`, `for`-loop etc.) or fixed.

I just got caught by this by defining a recursive `{% macro %}` inside a `{% block %}`:
```
{% block content %}

{% macro hello(name, counter=3) %}
  Hello {{name}}!
  {% if counter > 0 %}
    {{ hello(name, counter=counter-1) }}
  {% endif %}
{% endmacro %}

{{ hello('Malthe') }}
{% endblock %}
```

**Comment by ipmcc** [comment]
Also doesn't work inside a `with` block.

**Comment by jeffwidman** [comment]
PRs clarifying the docs are always welcome.

**Comment by mitsuhiko** [comment]
The original issues is fixed in 2.9 with the new identifier tracking. The block issue is unrelated and tracked in #603 now.
